File: web/J Voice web/src/store/featureFlags.js
/**
 * The app-level feature flags at `/flags` in the Realtime Database.
 *
 * These are the switches the Android app reads to decide which controls to show.
 * The console is the only place they are ever written; the rules allow the write
 * to a super admin alone, so a non-admin turning a switch here gets a permission
 * error from Firebase rather than a silent no-op.
 *
 * Kept out of the main store on purpose. Everything in the store is Firestore
 * content that flows through the reducer and the diff-based write-back; this is
 * four booleans in a different database with different rules, and pushing them
 * through that machinery would mean teaching it a second backend for no gain.
 */
import { onValue, ref, set, get } from 'firebase/database'
import { getFirebaseDb, isFirebaseReady } from '../firebase.js'

const NODE = 'flags'

/**
 * The flags the app understands, in the order they should be listed.
 *
 * The keys have to match `FeatureFlags.Keys` in the Android app character for
 * character — a typo here writes a key nothing reads, and the switch appears to
 * do nothing at all.
 */
export const FLAG_DEFS = [
  {
    key: 'newsLocationDropdown',
    label: 'Location picker',
    desc: 'The city dropdown beside the J Voice logo in the news toolbar'
  },
  {
    key: 'locationChips',
    label: 'Category chips',
    desc: 'The scrolling filter chips under the toolbar in the swipe feed'
  },
  {
    key: 'shortsTab',
    label: 'Clips flow',
    desc: 'Short-video clips: the reader bottom-bar tab and the screen behind it. ' +
      'Turning this off also redirects anyone currently on the Clips screen.'
  },
  {
    key: 'moduleIcon',
    label: 'Feed switch icon',
    desc: 'The toolbar icon left of search that swaps between the swipe and classic feeds'
  }
]

/**
 * A missing flag means ON.
 *
 * Matches `FeatureFlags.interpret` in the app, and matters here because the node
 * starts out empty: every switch must read as on before anybody has ever saved
 * one, or the console would claim the app is stripped bare when it is not.
 */
export function interpret(value) {
  if (value === undefined || value === null) return true
  if (typeof value === 'boolean') return value
  if (typeof value === 'number') return value !== 0
  if (typeof value === 'string') {
    const v = value.trim().toLowerCase()
    return !(v === 'false' || v === 'no' || v === '0')
  }
  return true
}

/** Fills in every known key so the UI never binds a switch to `undefined`. */
export function normalise(raw) {
  const source = raw ?? {}
  const out = {}
  for (const def of FLAG_DEFS) out[def.key] = interpret(source[def.key])
  return out
}

/** Subscribes to the live flags. Returns an unsubscribe, or a no-op when offline. */
export function watchFlags(onChange) {
  if (!isFirebaseReady()) {
    onChange(normalise(null))
    return () => {}
  }
  return onValue(
    ref(getFirebaseDb(), NODE),
    (snap) => onChange(normalise(snap.val())),
    // A rules failure must not leave the switches stuck mid-load. Showing the
    // defaults is honest: it is what the app itself would fall back to.
    () => onChange(normalise(null))
  )
}

/**
 * Writes one flag.
 *
 * Writes the single child rather than the whole object so two admins editing
 * different switches at the same moment do not overwrite each other. Rejects
 * (rather than resolving quietly) when the rules refuse, so the caller can say so.
 */
export async function setFlag(key, value) {
  if (!isFirebaseReady()) throw new Error('Firebase is not configured')
  await set(ref(getFirebaseDb(), `${NODE}/${key}`), Boolean(value))
}

/**
 * Writes any flags the node does not have yet, leaving existing ones alone.
 *
 * Used once from the console so the node exists and is visible in the Firebase
 * dashboard. Not required for correctness — the app treats an absent node as
 * everything-on — but an empty node is hard to tell apart from a broken one.
 */
export async function seedMissingFlags() {
  if (!isFirebaseReady()) throw new Error('Firebase is not configured')
  const snap = await get(ref(getFirebaseDb(), NODE))
  const current = snap.val() ?? {}
  const writes = FLAG_DEFS
    .filter((def) => current[def.key] === undefined)
    .map((def) => set(ref(getFirebaseDb(), `${NODE}/${def.key}`), true))
  await Promise.all(writes)
  return writes.length
}

File: android/app/src/main/java/com/jvoice/core/flags/FeatureFlags.kt
package com.jvoice.core.flags

import android.util.Log
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.ReadOnlyComposable
import androidx.compose.runtime.compositionLocalOf
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import com.jvoice.core.firebase.FirebaseAvailability
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * App-level feature flags, held at `/flags` in the Realtime Database.
 *
 * ## Why RTDB and not Firestore
 *
 * This is one small object that every launch reads and that has to take effect
 * *live* — the point of a flag is turning something off without shipping a build.
 * RTDB value events give that for the cost of one tiny node; a Firestore document
 * listener would work too but the session signals already live here, so the app
 * keeps one connection rather than opening a second for four booleans.
 *
 * ## Missing means ON
 *
 * Every flag defaults to `true` when the node is absent or unreadable. That is
 * deliberate and it is the direction that fails safely: a network blip, a deleted
 * node or a typo'd key leaves the app exactly as it shipped, rather than silently
 * stripping the toolbar and the tabs and looking broken. A feature is only ever
 * hidden by an *explicit* `false`.
 *
 * The same reasoning as [com.jvoice.core.auth.AuthGate.isEnabled], and the string
 * cases matter for the same reason: the Firebase console writes hand-typed values
 * as strings, so `"false"` is what an operator actually produces.
 */
object FeatureFlags {

    private const val TAG = "FeatureFlags"
    private const val NODE = "flags"

    /**
     * The flag keys, as stored. Named after what they control rather than where
     * they appear, so moving a control does not orphan its flag.
     */
    object Keys {
        /** The location picker in the news toolbar. */
        const val NEWS_LOCATION_DROPDOWN = "newsLocationDropdown"

        /** The row of location filter chips in the swipe feed. */
        const val LOCATION_CHIPS = "locationChips"

        /** The Clips / Shorts tab in the reader's bottom bar. */
        const val SHORTS_TAB = "shortsTab"

        /** The module-switch icon in the toolbar, left of search. */
        const val MODULE_ICON = "moduleIcon"
    }

    /** Current values. Absent keys are simply not present and read as ON. */
    private val _flags = MutableStateFlow<Map<String, Boolean>>(emptyMap())
    val flags: StateFlow<Map<String, Boolean>> = _flags.asStateFlow()

    private var listener: ValueEventListener? = null

    fun start() {
        if (listener != null || !FirebaseAvailability.isAvailable) return
        val ref = FirebaseDatabase.getInstance().reference.child(NODE)

        val l = object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                val parsed = snapshot.children.mapNotNull { child ->
                    val key = child.key ?: return@mapNotNull null
                    key to interpret(child.value)
                }.toMap()
                _flags.value = parsed
                Log.i(TAG, "flags: $parsed")
            }

            override fun onCancelled(error: DatabaseError) {
                // Leaves the map empty, which means every flag reads as ON - see
                // the class note. A rules problem must not blank the UI.
                Log.w(TAG, "flags unreadable, defaulting everything on: ${error.message}")
            }
        }
        ref.addValueEventListener(l)
        listener = l
    }

    /** Only an explicit falsy value turns a flag off. */
    private fun interpret(value: Any?): Boolean = when (value) {
        null -> true
        is Boolean -> value
        is Number -> value.toInt() != 0
        is String -> !(
            value.equals("false", ignoreCase = true) ||
                value.equals("no", ignoreCase = true) ||
                value == "0"
            )
        else -> true
    }

    /** Is [key] on? Unknown keys are on. */
    fun isEnabled(key: String): Boolean = _flags.value[key] ?: true
}

/**
 * The flags, readable from any composable without threading them through every
 * screen signature — the same approach
 * [com.jvoice.core.i18n.LocalAppLanguage] uses.
 *
 * `compositionLocalOf` rather than the static variant because these change at
 * runtime and every screen reading one has to recompose.
 */
val LocalFeatureFlags = compositionLocalOf { emptyMap<String, Boolean>() }

/** Reads one flag inside a composable. Unknown keys are on. */
@Composable
@ReadOnlyComposable
fun flagEnabled(key: String): Boolean = LocalFeatureFlags.current[key] ?: true

/**
 * Puts a whole destination behind a flag, not just the control that opens it.
 *
 * [flagEnabled] hides an *entrance* - a tab, a toolbar icon - which is all a
 * control needs. A flow needs more, because hiding the way in does nothing about
 * whoever is already inside:
 *
 *  * a reader sitting on the screen when an admin flips the flag keeps the
 *    screen, and only the tab underneath it vanishes;
 *  * Navigation restores the last route after process death, so an app killed on
 *    a flagged-off screen reopens straight back onto it.
 *
 * One live check covers both. [content] composes only while the flag is on, and
 * [onBlocked] - a redirect to somewhere always reachable - runs the moment it is
 * not: on first composition for the restored-route case, and on recomposition
 * for the live flip.
 *
 * Nothing is drawn in the blocked state on purpose. The redirect is immediate,
 * and a flash of "this feature is off" on the way out reads as a crash rather
 * than a decision.
 */
@Composable
fun FlaggedRoute(key: String, onBlocked: () -> Unit, content: @Composable () -> Unit) {
    val enabled = flagEnabled(key)
    LaunchedEffect(enabled) { if (!enabled) onBlocked() }
    if (enabled) content()
}
